platforms/software/testing-base/src/integTest/groovy/org/gradle/testing/TestEventServicesIntegrationTest.groovy

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.testing
+
+import org.gradle.api.internal.tasks.testing.TestDescriptorInternal
+import org.gradle.api.internal.tasks.testing.operations.ExecuteTestBuildOperationType
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.BuildOperationsFixture
+import org.gradle.integtests.fixtures.TestResources
+import org.junit.Rule
+
+class TestEventServicesIntegrationTest extends AbstractIntegrationSpec {
+
+    @Rule
+    final TestResources resources = new TestResources(testDirectoryProvider)
+
+    def operations = new BuildOperationsFixture(executer, temporaryFolder)
+
+    def "emits build operations for custom test"() {
+        given:
+        buildFile("""
+            import java.time.Instant
+
+            abstract class CustomTestTask extends DefaultTask {
+                @Inject
+                abstract TestEventService getTestEventService()
+
+                @TaskAction
+                void runTests() {
+                   try (def generator = getTestEventService().generateTestEvents("Custom test root")) {
+                       generator.started(Instant.now())
+                       try (def mySuite = generator.createCompositeNested("My Suite")) {
+                            mySuite.started(Instant.now())
+                            try (def myTest = mySuite.createSolitaryNested("MyTestInternal", "My test!")) {
+                                 myTest.started(Instant.now())
```

Comment:
Could you demonstrate some other structures here?

e.g., does it make sense to have multiple started/completed calls within one createSolitaryNested?